// esa-cli/dist/commands/lang.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import SelectItems from '../components/selectInput.js';
import t from '../i18n/index.js';
import logger from '../libs/logger.js';
import { updateCliConfigFile } from '../utils/fileUtils/index.js';
const docs = {
    command: 'lang',
    describe: `🌐 ${t('lang_describe').d('Set the language of the CLI')}`,
    handler: () => __awaiter(void 0, void 0, void 0, function* () {
        SelectItems({
            items: [
                { label: 'English', value: 'en' },
                { label: '简体中文', value: 'zh_CN' }
            ],
            handleSelect: (item) => __awaiter(void 0, void 0, void 0, function* () {
                yield updateCliConfigFile({
                    lang: item.value
                });
                logger.success(t('lang_success').d('Language switched'));
            })
        });
    })
};
export default docs;
