// esa-cli/dist/commands/logout.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import t from '../i18n/index.js';
import logger from '../libs/logger.js';
import { getCliConfig, updateCliConfigFile } from '../utils/fileUtils/index.js';
const logout = {
    command: 'logout',
    describe: `🚪 ${t('logout_describe').d('Logout')}`,
    builder: (yargs) => {
        return yargs;
    },
    handler: () => {
        handleLogout();
    }
};
export default logout;
export function handleLogout() {
    return __awaiter(this, void 0, void 0, function* () {
        let cliConfig = getCliConfig();
        if (!cliConfig) {
            return;
        }
        if (!cliConfig.auth) {
            cliConfig.auth = {
                accessKeyId: '',
                accessKeySecret: '',
                securityToken: ''
            };
        }
        else {
            cliConfig.auth.accessKeyId = '';
            cliConfig.auth.accessKeySecret = '';
            if ('securityToken' in cliConfig.auth) {
                cliConfig.auth.securityToken = '';
            }
        }
        yield updateCliConfigFile(cliConfig);
        logger.success(t('logout_success').d('Logout successfully'));
    });
}
