// esa-cli/dist/commands/routine/delete.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import t from '../../i18n/index.js';
import { ApiService } from '../../libs/apiService.js';
import logger from '../../libs/logger.js';
import { checkIsLoginSuccess } from '../utils.js';
const deleteCommand = {
    command: 'delete <projectName>',
    aliases: ['delete <routineName>'],
    describe: `🗑  ${t('delete_describe').d('Delete a project')}`,
    builder: (yargs) => {
        return yargs
            .positional('projectName', {
            describe: t('delete_routineName_positional_describe').d('The name of the project to delete'),
            type: 'string',
            array: true,
            demandOption: true
        })
            .usage(`${t('common_usage').d('Usage')}: $0 delete <projectName>`);
    },
    handler: (argv) => __awaiter(void 0, void 0, void 0, function* () {
        handleDelete(argv);
    })
};
export default deleteCommand;
export function handleDelete(argv) {
    return __awaiter(this, void 0, void 0, function* () {
        const isSuccess = yield checkIsLoginSuccess();
        if (!isSuccess)
            return;
        const routineName = argv.projectName || argv.routineName;
        const req = { Name: routineName };
        return yield deleteRoutineFromUserAccount(req);
    });
}
export function deleteRoutineFromUserAccount(req) {
    return __awaiter(this, void 0, void 0, function* () {
        const server = yield ApiService.getInstance();
        const res = yield server.deleteRoutine(req);
        if ((res === null || res === void 0 ? void 0 : res.Status) === 'OK') {
            logger.success(t('delete_success').d('Delete success!'));
        }
        else {
            logger.error(t('delete_error').d('Delete error!'));
        }
    });
}
