// esa-cli/dist/commands/deployments/delete.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import chalk from 'chalk';
import { displayMultiSelectTable } from '../../components/mutiSelectTable.js';
import t from '../../i18n/index.js';
import { ApiService } from '../../libs/apiService.js';
import logger from '../../libs/logger.js';
import { validRoutine } from '../../utils/checkIsRoutineCreated.js';
import { getProjectConfig } from '../../utils/fileUtils/index.js';
import { checkDirectory, checkIsLoginSuccess, getRoutineCodeVersions } from '../utils.js';
const deploymentsDelete = {
    command: 'delete [deploymentId..]',
    describe: `🗑  ${t('deployments_delete_describe').d('Delete one or more deployment versions')}`,
    builder: (yargs) => {
        return yargs
            .positional('deploymentId', {
            describe: t('deployments_delete_id_describe').d('The ID of the deployments to delete'),
            type: 'string',
            demandOption: true
        })
            .option('i', {
            describe: t('delete_interactive_mode').d('Delete deployments by using interactive mode'),
            type: 'boolean'
        });
    },
    handler: (argv) => __awaiter(void 0, void 0, void 0, function* () {
        handleDeleteDeployments(argv);
    })
};
export default deploymentsDelete;
export function handleDeleteDeployments(argv) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!checkDirectory()) {
            return;
        }
        const projectConfig = getProjectConfig();
        if (!projectConfig)
            return logger.notInProject();
        const isSuccess = yield checkIsLoginSuccess();
        if (!isSuccess)
            return;
        yield validRoutine(projectConfig.name);
        const server = yield ApiService.getInstance();
        let versions = argv.deploymentId;
        const isInteractive = argv.i;
        if (isInteractive) {
            const { allVersions, stagingVersions, productionVersions } = yield getRoutineCodeVersions(projectConfig.name);
            // Show information about versions being deployed
            if (stagingVersions.length > 0 || productionVersions.length > 0) {
                logger.log(chalk.yellow('⚠️  Currently deploying versions:'));
                if (stagingVersions.length > 0) {
                    logger.log(chalk.yellow(`   Staging: ${stagingVersions.join(',')}`));
                }
                if (productionVersions.length > 0) {
                    logger.log(chalk.yellow(`   Production: ${productionVersions.join(',')}`));
                }
                logger.log('');
            }
            logger.log(t('delete_deployments_table_title').d('  Version ID            Description'));
            // Filter out versions being deployed
            const selectList = allVersions
                .filter((item) => {
                var _a, _b;
                if (stagingVersions.length === 0 && productionVersions.length === 0)
                    return true;
                return (!stagingVersions.includes((_a = item.codeVersion) !== null && _a !== void 0 ? _a : '') &&
                    !productionVersions.includes((_b = item.codeVersion) !== null && _b !== void 0 ? _b : ''));
            })
                .map((item) => {
                return {
                    label: item.codeVersion + '   ' + item.codeDescription
                };
            });
            if (selectList.length === 0) {
                logger.error('No deletable versions found. All versions are currently deployed.');
                return;
            }
            versions = (yield displayMultiSelectTable(selectList, 1, 100)).map((item) => item.slice(0, item.indexOf(' ')));
        }
        for (let i = 0; i < versions.length; i++) {
            const version = versions[i];
            const req = {
                Name: projectConfig.name,
                CodeVersion: version
            };
            const res = yield server.deleteRoutineCodeVersion(req);
            if ((res === null || res === void 0 ? void 0 : res.Status) === 'OK') {
                logger.success(`${t('deployments_delete_success').d('Delete success')}: ${version}`);
            }
            else {
                logger.error(`🙅 ${t('deployments_delete_failed').d('Delete failed')}: ${version}`);
            }
        }
    });
}
