// @alicloud/esa20240910/dist/models/CreateHttpsBasicConfigurationRequest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateHttpsBasicConfigurationRequest = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
class CreateHttpsBasicConfigurationRequest extends $dara.Model {
    static names() {
        return {
            ciphersuite: 'Ciphersuite',
            ciphersuiteGroup: 'CiphersuiteGroup',
            http2: 'Http2',
            http3: 'Http3',
            https: 'Https',
            ocspStapling: 'OcspStapling',
            rule: 'Rule',
            ruleEnable: 'RuleEnable',
            ruleName: 'RuleName',
            sequence: 'Sequence',
            siteId: 'SiteId',
            tls10: 'Tls10',
            tls11: 'Tls11',
            tls12: 'Tls12',
            tls13: 'Tls13',
        };
    }
    static types() {
        return {
            ciphersuite: 'string',
            ciphersuiteGroup: 'string',
            http2: 'string',
            http3: 'string',
            https: 'string',
            ocspStapling: 'string',
            rule: 'string',
            ruleEnable: 'string',
            ruleName: 'string',
            sequence: 'number',
            siteId: 'number',
            tls10: 'string',
            tls11: 'string',
            tls12: 'string',
            tls13: 'string',
        };
    }
    validate() {
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.CreateHttpsBasicConfigurationRequest = CreateHttpsBasicConfigurationRequest;
