// @alicloud/esa20240910/dist/models/GetEdgeContainerAppStatusResponseBodyAppStatus.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetEdgeContainerAppStatusResponseBodyAppStatus = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const GetEdgeContainerAppStatusResponseBodyAppStatusRegions_1 = require("./GetEdgeContainerAppStatusResponseBodyAppStatusRegions");
class GetEdgeContainerAppStatusResponseBodyAppStatus extends $dara.Model {
    static names() {
        return {
            baseLineVersion: 'BaseLineVersion',
            deployStatus: 'DeployStatus',
            deployTime: 'DeployTime',
            deployedVersion: 'DeployedVersion',
            expectPercentage: 'ExpectPercentage',
            fullRelease: 'FullRelease',
            publishEnv: 'PublishEnv',
            publishPercentage: 'PublishPercentage',
            publishStatus: 'PublishStatus',
            publishTime: 'PublishTime',
            publishType: 'PublishType',
            publishingVersion: 'PublishingVersion',
            regions: 'Regions',
            rollbackTime: 'RollbackTime',
            unDeployTime: 'UnDeployTime',
        };
    }
    static types() {
        return {
            baseLineVersion: 'string',
            deployStatus: 'string',
            deployTime: 'string',
            deployedVersion: 'string',
            expectPercentage: 'number',
            fullRelease: 'boolean',
            publishEnv: 'string',
            publishPercentage: 'number',
            publishStatus: 'string',
            publishTime: 'string',
            publishType: 'string',
            publishingVersion: 'string',
            regions: GetEdgeContainerAppStatusResponseBodyAppStatusRegions_1.GetEdgeContainerAppStatusResponseBodyAppStatusRegions,
            rollbackTime: 'string',
            unDeployTime: 'string',
        };
    }
    validate() {
        if (this.regions && typeof this.regions.validate === 'function') {
            this.regions.validate();
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.GetEdgeContainerAppStatusResponseBodyAppStatus = GetEdgeContainerAppStatusResponseBodyAppStatus;
