// @alicloud/esa20240910/dist/models/GetOriginProtectionResponseBodyDiffIpwhitelist.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetOriginProtectionResponseBodyDiffIPWhitelist = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const GetOriginProtectionResponseBodyDiffIpwhitelistAddedIpwhitelist_1 = require("./GetOriginProtectionResponseBodyDiffIpwhitelistAddedIpwhitelist");
const GetOriginProtectionResponseBodyDiffIpwhitelistNoChangeIpWhitelist_1 = require("./GetOriginProtectionResponseBodyDiffIpwhitelistNoChangeIpWhitelist");
const GetOriginProtectionResponseBodyDiffIpwhitelistRemovedIpwhitelist_1 = require("./GetOriginProtectionResponseBodyDiffIpwhitelistRemovedIpwhitelist");
class GetOriginProtectionResponseBodyDiffIPWhitelist extends $dara.Model {
    static names() {
        return {
            addedIPWhitelist: 'AddedIPWhitelist',
            noChangeIpWhitelist: 'NoChangeIpWhitelist',
            removedIPWhitelist: 'RemovedIPWhitelist',
        };
    }
    static types() {
        return {
            addedIPWhitelist: GetOriginProtectionResponseBodyDiffIpwhitelistAddedIpwhitelist_1.GetOriginProtectionResponseBodyDiffIPWhitelistAddedIPWhitelist,
            noChangeIpWhitelist: GetOriginProtectionResponseBodyDiffIpwhitelistNoChangeIpWhitelist_1.GetOriginProtectionResponseBodyDiffIPWhitelistNoChangeIpWhitelist,
            removedIPWhitelist: GetOriginProtectionResponseBodyDiffIpwhitelistRemovedIpwhitelist_1.GetOriginProtectionResponseBodyDiffIPWhitelistRemovedIPWhitelist,
        };
    }
    validate() {
        if (this.addedIPWhitelist && typeof this.addedIPWhitelist.validate === 'function') {
            this.addedIPWhitelist.validate();
        }
        if (this.noChangeIpWhitelist && typeof this.noChangeIpWhitelist.validate === 'function') {
            this.noChangeIpWhitelist.validate();
        }
        if (this.removedIPWhitelist && typeof this.removedIPWhitelist.validate === 'function') {
            this.removedIPWhitelist.validate();
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.GetOriginProtectionResponseBodyDiffIPWhitelist = GetOriginProtectionResponseBodyDiffIPWhitelist;
