// @alicloud/esa20240910/dist/models/GetRecordResponseBodyRecordModel.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetRecordResponseBodyRecordModel = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const GetRecordResponseBodyRecordModelAuthConf_1 = require("./GetRecordResponseBodyRecordModelAuthConf");
const GetRecordResponseBodyRecordModelData_1 = require("./GetRecordResponseBodyRecordModelData");
class GetRecordResponseBodyRecordModel extends $dara.Model {
    static names() {
        return {
            authConf: 'AuthConf',
            bizName: 'BizName',
            comment: 'Comment',
            createTime: 'CreateTime',
            data: 'Data',
            hostPolicy: 'HostPolicy',
            proxied: 'Proxied',
            recordCname: 'RecordCname',
            recordId: 'RecordId',
            recordName: 'RecordName',
            recordSourceType: 'RecordSourceType',
            recordType: 'RecordType',
            siteId: 'SiteId',
            siteName: 'SiteName',
            ttl: 'Ttl',
            updateTime: 'UpdateTime',
        };
    }
    static types() {
        return {
            authConf: GetRecordResponseBodyRecordModelAuthConf_1.GetRecordResponseBodyRecordModelAuthConf,
            bizName: 'string',
            comment: 'string',
            createTime: 'string',
            data: GetRecordResponseBodyRecordModelData_1.GetRecordResponseBodyRecordModelData,
            hostPolicy: 'string',
            proxied: 'boolean',
            recordCname: 'string',
            recordId: 'number',
            recordName: 'string',
            recordSourceType: 'string',
            recordType: 'string',
            siteId: 'number',
            siteName: 'string',
            ttl: 'number',
            updateTime: 'string',
        };
    }
    validate() {
        if (this.authConf && typeof this.authConf.validate === 'function') {
            this.authConf.validate();
        }
        if (this.data && typeof this.data.validate === 'function') {
            this.data.validate();
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.GetRecordResponseBodyRecordModel = GetRecordResponseBodyRecordModel;
