// @alicloud/esa20240910/dist/models/GetSiteResponseBodySiteModel.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetSiteResponseBodySiteModel = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
class GetSiteResponseBodySiteModel extends $dara.Model {
    static names() {
        return {
            accessType: 'AccessType',
            cnameZone: 'CnameZone',
            coverage: 'Coverage',
            createTime: 'CreateTime',
            instanceId: 'InstanceId',
            nameServerList: 'NameServerList',
            offlineReason: 'OfflineReason',
            planName: 'PlanName',
            planSpecName: 'PlanSpecName',
            resourceGroupId: 'ResourceGroupId',
            siteId: 'SiteId',
            siteName: 'SiteName',
            status: 'Status',
            tags: 'Tags',
            updateTime: 'UpdateTime',
            vanityNSList: 'VanityNSList',
            verifyCode: 'VerifyCode',
            versionManagement: 'VersionManagement',
        };
    }
    static types() {
        return {
            accessType: 'string',
            cnameZone: 'string',
            coverage: 'string',
            createTime: 'string',
            instanceId: 'string',
            nameServerList: 'string',
            offlineReason: 'string',
            planName: 'string',
            planSpecName: 'string',
            resourceGroupId: 'string',
            siteId: 'number',
            siteName: 'string',
            status: 'string',
            tags: { 'type': 'map', 'keyType': 'string', 'valueType': 'any' },
            updateTime: 'string',
            vanityNSList: { 'type': 'map', 'keyType': 'string', 'valueType': 'string' },
            verifyCode: 'string',
            versionManagement: 'boolean',
        };
    }
    validate() {
        if (this.tags) {
            $dara.Model.validateMap(this.tags);
        }
        if (this.vanityNSList) {
            $dara.Model.validateMap(this.vanityNSList);
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.GetSiteResponseBodySiteModel = GetSiteResponseBodySiteModel;
