// @alicloud/esa20240910/dist/models/GetWafFilterResponseBodyFilterFields.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetWafFilterResponseBodyFilterFields = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const GetWafFilterResponseBodyFilterFieldsLogics_1 = require("./GetWafFilterResponseBodyFilterFieldsLogics");
const GetWafFilterResponseBodyFilterFieldsSelector_1 = require("./GetWafFilterResponseBodyFilterFieldsSelector");
class GetWafFilterResponseBodyFilterFields extends $dara.Model {
    static names() {
        return {
            enable: 'Enable',
            key: 'Key',
            label: 'Label',
            logics: 'Logics',
            minPlan: 'MinPlan',
            selector: 'Selector',
            sub: 'Sub',
            subTip: 'SubTip',
        };
    }
    static types() {
        return {
            enable: 'boolean',
            key: 'string',
            label: 'string',
            logics: { 'type': 'array', 'itemType': GetWafFilterResponseBodyFilterFieldsLogics_1.GetWafFilterResponseBodyFilterFieldsLogics },
            minPlan: 'string',
            selector: GetWafFilterResponseBodyFilterFieldsSelector_1.GetWafFilterResponseBodyFilterFieldsSelector,
            sub: 'boolean',
            subTip: 'string',
        };
    }
    validate() {
        if (Array.isArray(this.logics)) {
            $dara.Model.validateArray(this.logics);
        }
        if (this.selector && typeof this.selector.validate === 'function') {
            this.selector.validate();
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.GetWafFilterResponseBodyFilterFields = GetWafFilterResponseBodyFilterFields;
