// @alicloud/esa20240910/dist/models/ListCertificatesResponseBodyResult.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListCertificatesResponseBodyResult = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const ListCertificatesResponseBodyResultDcv_1 = require("./ListCertificatesResponseBodyResultDcv");
class ListCertificatesResponseBodyResult extends $dara.Model {
    static names() {
        return {
            applyCode: 'ApplyCode',
            applyMessage: 'ApplyMessage',
            casId: 'CasId',
            commonName: 'CommonName',
            createTime: 'CreateTime',
            DCV: 'DCV',
            fingerprintSha256: 'FingerprintSha256',
            id: 'Id',
            issuer: 'Issuer',
            issuerCN: 'IssuerCN',
            name: 'Name',
            notAfter: 'NotAfter',
            notBefore: 'NotBefore',
            pubAlg: 'PubAlg',
            region: 'Region',
            SAN: 'SAN',
            serialNumber: 'SerialNumber',
            sigAlg: 'SigAlg',
            status: 'Status',
            type: 'Type',
            updateTime: 'UpdateTime',
        };
    }
    static types() {
        return {
            applyCode: 'number',
            applyMessage: 'string',
            casId: 'string',
            commonName: 'string',
            createTime: 'string',
            DCV: { 'type': 'array', 'itemType': ListCertificatesResponseBodyResultDcv_1.ListCertificatesResponseBodyResultDCV },
            fingerprintSha256: 'string',
            id: 'string',
            issuer: 'string',
            issuerCN: 'string',
            name: 'string',
            notAfter: 'string',
            notBefore: 'string',
            pubAlg: 'string',
            region: 'string',
            SAN: 'string',
            serialNumber: 'string',
            sigAlg: 'string',
            status: 'string',
            type: 'string',
            updateTime: 'string',
        };
    }
    validate() {
        if (Array.isArray(this.DCV)) {
            $dara.Model.validateArray(this.DCV);
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.ListCertificatesResponseBodyResult = ListCertificatesResponseBodyResult;
