// @alicloud/esa20240910/dist/models/ListWafRulesResponseBodyRules.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListWafRulesResponseBodyRules = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const WafRuleConfig_1 = require("./WafRuleConfig");
const WafTimer_1 = require("./WafTimer");
class ListWafRulesResponseBodyRules extends $dara.Model {
    static names() {
        return {
            action: 'Action',
            characteristicsFields: 'CharacteristicsFields',
            config: 'Config',
            fields: 'Fields',
            id: 'Id',
            name: 'Name',
            phase: 'Phase',
            position: 'Position',
            rulesetId: 'RulesetId',
            skip: 'Skip',
            status: 'Status',
            tags: 'Tags',
            timer: 'Timer',
            type: 'Type',
            updateTime: 'UpdateTime',
        };
    }
    static types() {
        return {
            action: 'string',
            characteristicsFields: { 'type': 'array', 'itemType': 'string' },
            config: WafRuleConfig_1.WafRuleConfig,
            fields: { 'type': 'array', 'itemType': 'string' },
            id: 'number',
            name: 'string',
            phase: 'string',
            position: 'number',
            rulesetId: 'number',
            skip: 'string',
            status: 'string',
            tags: { 'type': 'array', 'itemType': 'string' },
            timer: WafTimer_1.WafTimer,
            type: 'string',
            updateTime: 'string',
        };
    }
    validate() {
        if (Array.isArray(this.characteristicsFields)) {
            $dara.Model.validateArray(this.characteristicsFields);
        }
        if (this.config && typeof this.config.validate === 'function') {
            this.config.validate();
        }
        if (Array.isArray(this.fields)) {
            $dara.Model.validateArray(this.fields);
        }
        if (Array.isArray(this.tags)) {
            $dara.Model.validateArray(this.tags);
        }
        if (this.timer && typeof this.timer.validate === 'function') {
            this.timer.validate();
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.ListWafRulesResponseBodyRules = ListWafRulesResponseBodyRules;
