// @alicloud/esa20240910/dist/models/ListUserRatePlanInstancesResponseBodyInstanceInfo.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListUserRatePlanInstancesResponseBodyInstanceInfo = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const ListUserRatePlanInstancesResponseBodyInstanceInfoSites_1 = require("./ListUserRatePlanInstancesResponseBodyInstanceInfoSites");
class ListUserRatePlanInstancesResponseBodyInstanceInfo extends $dara.Model {
    static names() {
        return {
            billingMode: 'BillingMode',
            botInstanceLevel: 'BotInstanceLevel',
            botRequest: 'BotRequest',
            coverages: 'Coverages',
            createTime: 'CreateTime',
            crossborderTraffic: 'CrossborderTraffic',
            ddosBurstableDomesticProtection: 'DdosBurstableDomesticProtection',
            ddosBurstableOverseasProtection: 'DdosBurstableOverseasProtection',
            ddosInstanceLevel: 'DdosInstanceLevel',
            duration: 'Duration',
            edgeRoutineRquest: 'EdgeRoutineRquest',
            edgeWafRequest: 'EdgeWafRequest',
            expireTime: 'ExpireTime',
            instanceId: 'InstanceId',
            layer4Traffic: 'Layer4Traffic',
            layer4TrafficIntl: 'Layer4TrafficIntl',
            planName: 'PlanName',
            planTraffic: 'PlanTraffic',
            planType: 'PlanType',
            siteQuota: 'SiteQuota',
            sites: 'Sites',
            smartRoutingRequest: 'SmartRoutingRequest',
            staticRequest: 'StaticRequest',
            status: 'Status',
            subscribeType: 'SubscribeType',
        };
    }
    static types() {
        return {
            billingMode: 'string',
            botInstanceLevel: 'string',
            botRequest: 'string',
            coverages: 'string',
            createTime: 'string',
            crossborderTraffic: 'string',
            ddosBurstableDomesticProtection: 'string',
            ddosBurstableOverseasProtection: 'string',
            ddosInstanceLevel: 'string',
            duration: 'number',
            edgeRoutineRquest: 'string',
            edgeWafRequest: 'string',
            expireTime: 'string',
            instanceId: 'string',
            layer4Traffic: 'string',
            layer4TrafficIntl: 'string',
            planName: 'string',
            planTraffic: 'string',
            planType: 'string',
            siteQuota: 'string',
            sites: { 'type': 'array', 'itemType': ListUserRatePlanInstancesResponseBodyInstanceInfoSites_1.ListUserRatePlanInstancesResponseBodyInstanceInfoSites },
            smartRoutingRequest: 'string',
            staticRequest: 'string',
            status: 'string',
            subscribeType: 'string',
        };
    }
    validate() {
        if (Array.isArray(this.sites)) {
            $dara.Model.validateArray(this.sites);
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.ListUserRatePlanInstancesResponseBodyInstanceInfo = ListUserRatePlanInstancesResponseBodyInstanceInfo;
