// @alicloud/esa20240910/dist/models/PurgeCachesRequestContent.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PurgeCachesRequestContent = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const PurgeCachesRequestContentCacheKeys_1 = require("./PurgeCachesRequestContentCacheKeys");
class PurgeCachesRequestContent extends $dara.Model {
    static names() {
        return {
            cacheKeys: 'CacheKeys',
            cacheTags: 'CacheTags',
            directories: 'Directories',
            files: 'Files',
            hostnames: 'Hostnames',
            ignoreParams: 'IgnoreParams',
            purgeAll: 'PurgeAll',
        };
    }
    static types() {
        return {
            cacheKeys: { 'type': 'array', 'itemType': PurgeCachesRequestContentCacheKeys_1.PurgeCachesRequestContentCacheKeys },
            cacheTags: { 'type': 'array', 'itemType': 'string' },
            directories: { 'type': 'array', 'itemType': 'string' },
            files: { 'type': 'array', 'itemType': 'any' },
            hostnames: { 'type': 'array', 'itemType': 'string' },
            ignoreParams: { 'type': 'array', 'itemType': 'string' },
            purgeAll: 'boolean',
        };
    }
    validate() {
        if (Array.isArray(this.cacheKeys)) {
            $dara.Model.validateArray(this.cacheKeys);
        }
        if (Array.isArray(this.cacheTags)) {
            $dara.Model.validateArray(this.cacheTags);
        }
        if (Array.isArray(this.directories)) {
            $dara.Model.validateArray(this.directories);
        }
        if (Array.isArray(this.files)) {
            $dara.Model.validateArray(this.files);
        }
        if (Array.isArray(this.hostnames)) {
            $dara.Model.validateArray(this.hostnames);
        }
        if (Array.isArray(this.ignoreParams)) {
            $dara.Model.validateArray(this.ignoreParams);
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.PurgeCachesRequestContent = PurgeCachesRequestContent;
