// @alicloud/esa20240910/dist/models/UpdateLoadBalancerRequest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateLoadBalancerRequest = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const UpdateLoadBalancerRequestAdaptiveRouting_1 = require("./UpdateLoadBalancerRequestAdaptiveRouting");
const UpdateLoadBalancerRequestMonitor_1 = require("./UpdateLoadBalancerRequestMonitor");
const UpdateLoadBalancerRequestRandomSteering_1 = require("./UpdateLoadBalancerRequestRandomSteering");
const UpdateLoadBalancerRequestRules_1 = require("./UpdateLoadBalancerRequestRules");
class UpdateLoadBalancerRequest extends $dara.Model {
    static names() {
        return {
            adaptiveRouting: 'AdaptiveRouting',
            defaultPools: 'DefaultPools',
            description: 'Description',
            enabled: 'Enabled',
            fallbackPool: 'FallbackPool',
            id: 'Id',
            monitor: 'Monitor',
            randomSteering: 'RandomSteering',
            regionPools: 'RegionPools',
            rules: 'Rules',
            sessionAffinity: 'SessionAffinity',
            siteId: 'SiteId',
            steeringPolicy: 'SteeringPolicy',
            subRegionPools: 'SubRegionPools',
            ttl: 'Ttl',
        };
    }
    static types() {
        return {
            adaptiveRouting: UpdateLoadBalancerRequestAdaptiveRouting_1.UpdateLoadBalancerRequestAdaptiveRouting,
            defaultPools: { 'type': 'array', 'itemType': 'number' },
            description: 'string',
            enabled: 'boolean',
            fallbackPool: 'number',
            id: 'number',
            monitor: UpdateLoadBalancerRequestMonitor_1.UpdateLoadBalancerRequestMonitor,
            randomSteering: UpdateLoadBalancerRequestRandomSteering_1.UpdateLoadBalancerRequestRandomSteering,
            regionPools: 'any',
            rules: { 'type': 'array', 'itemType': UpdateLoadBalancerRequestRules_1.UpdateLoadBalancerRequestRules },
            sessionAffinity: 'string',
            siteId: 'number',
            steeringPolicy: 'string',
            subRegionPools: 'any',
            ttl: 'number',
        };
    }
    validate() {
        if (this.adaptiveRouting && typeof this.adaptiveRouting.validate === 'function') {
            this.adaptiveRouting.validate();
        }
        if (Array.isArray(this.defaultPools)) {
            $dara.Model.validateArray(this.defaultPools);
        }
        if (this.monitor && typeof this.monitor.validate === 'function') {
            this.monitor.validate();
        }
        if (this.randomSteering && typeof this.randomSteering.validate === 'function') {
            this.randomSteering.validate();
        }
        if (Array.isArray(this.rules)) {
            $dara.Model.validateArray(this.rules);
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.UpdateLoadBalancerRequest = UpdateLoadBalancerRequest;
