// @alicloud/esa20240910/dist/models/UpdateOriginRuleRequest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateOriginRuleRequest = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
class UpdateOriginRuleRequest extends $dara.Model {
    static names() {
        return {
            configId: 'ConfigId',
            dnsRecord: 'DnsRecord',
            follow302Enable: 'Follow302Enable',
            follow302MaxTries: 'Follow302MaxTries',
            follow302RetainArgs: 'Follow302RetainArgs',
            follow302RetainHeader: 'Follow302RetainHeader',
            follow302TargetHost: 'Follow302TargetHost',
            originHost: 'OriginHost',
            originHttpPort: 'OriginHttpPort',
            originHttpsPort: 'OriginHttpsPort',
            originMtls: 'OriginMtls',
            originReadTimeout: 'OriginReadTimeout',
            originScheme: 'OriginScheme',
            originSni: 'OriginSni',
            originVerify: 'OriginVerify',
            range: 'Range',
            rangeChunkSize: 'RangeChunkSize',
            rule: 'Rule',
            ruleEnable: 'RuleEnable',
            ruleName: 'RuleName',
            sequence: 'Sequence',
            siteId: 'SiteId',
        };
    }
    static types() {
        return {
            configId: 'number',
            dnsRecord: 'string',
            follow302Enable: 'string',
            follow302MaxTries: 'string',
            follow302RetainArgs: 'string',
            follow302RetainHeader: 'string',
            follow302TargetHost: 'string',
            originHost: 'string',
            originHttpPort: 'string',
            originHttpsPort: 'string',
            originMtls: 'string',
            originReadTimeout: 'string',
            originScheme: 'string',
            originSni: 'string',
            originVerify: 'string',
            range: 'string',
            rangeChunkSize: 'string',
            rule: 'string',
            ruleEnable: 'string',
            ruleName: 'string',
            sequence: 'number',
            siteId: 'number',
        };
    }
    validate() {
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.UpdateOriginRuleRequest = UpdateOriginRuleRequest;
