// @alicloud/esa20240910/dist/models/UpdateWaitingRoomEventRequest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateWaitingRoomEventRequest = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
class UpdateWaitingRoomEventRequest extends $dara.Model {
    static names() {
        return {
            customPageHtml: 'CustomPageHtml',
            description: 'Description',
            disableSessionRenewalEnable: 'DisableSessionRenewalEnable',
            enable: 'Enable',
            endTime: 'EndTime',
            jsonResponseEnable: 'JsonResponseEnable',
            language: 'Language',
            name: 'Name',
            newUsersPerMinute: 'NewUsersPerMinute',
            preQueueEnable: 'PreQueueEnable',
            preQueueStartTime: 'PreQueueStartTime',
            queuingMethod: 'QueuingMethod',
            queuingStatusCode: 'QueuingStatusCode',
            randomPreQueueEnable: 'RandomPreQueueEnable',
            sessionDuration: 'SessionDuration',
            siteId: 'SiteId',
            startTime: 'StartTime',
            totalActiveUsers: 'TotalActiveUsers',
            waitingRoomEventId: 'WaitingRoomEventId',
            waitingRoomType: 'WaitingRoomType',
        };
    }
    static types() {
        return {
            customPageHtml: 'string',
            description: 'string',
            disableSessionRenewalEnable: 'string',
            enable: 'string',
            endTime: 'string',
            jsonResponseEnable: 'string',
            language: 'string',
            name: 'string',
            newUsersPerMinute: 'string',
            preQueueEnable: 'string',
            preQueueStartTime: 'string',
            queuingMethod: 'string',
            queuingStatusCode: 'string',
            randomPreQueueEnable: 'string',
            sessionDuration: 'string',
            siteId: 'number',
            startTime: 'string',
            totalActiveUsers: 'string',
            waitingRoomEventId: 'number',
            waitingRoomType: 'string',
        };
    }
    validate() {
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.UpdateWaitingRoomEventRequest = UpdateWaitingRoomEventRequest;
