// @alicloud/esa20240910/dist/models/WafRuleConfig.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WafRuleConfig = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const WafRuleConfigActions_1 = require("./WafRuleConfigActions");
const WafRuleConfigAppPackage_1 = require("./WafRuleConfigAppPackage");
const WafRuleConfigAppSdk_1 = require("./WafRuleConfigAppSdk");
const WafRuleConfigManagedRulesets_1 = require("./WafRuleConfigManagedRulesets");
const WafRuleConfigRateLimit_1 = require("./WafRuleConfigRateLimit");
const WafRuleConfigSecurityLevel_1 = require("./WafRuleConfigSecurityLevel");
const WafTimer_1 = require("./WafTimer");
class WafRuleConfig extends $dara.Model {
    static names() {
        return {
            action: 'Action',
            actions: 'Actions',
            appPackage: 'AppPackage',
            appSdk: 'AppSdk',
            expression: 'Expression',
            id: 'Id',
            managedGroupId: 'ManagedGroupId',
            managedList: 'ManagedList',
            managedRulesets: 'ManagedRulesets',
            name: 'Name',
            notes: 'Notes',
            rateLimit: 'RateLimit',
            securityLevel: 'SecurityLevel',
            sigchl: 'Sigchl',
            status: 'Status',
            timer: 'Timer',
            type: 'Type',
            value: 'Value',
        };
    }
    static types() {
        return {
            action: 'string',
            actions: WafRuleConfigActions_1.WafRuleConfigActions,
            appPackage: WafRuleConfigAppPackage_1.WafRuleConfigAppPackage,
            appSdk: WafRuleConfigAppSdk_1.WafRuleConfigAppSdk,
            expression: 'string',
            id: 'number',
            managedGroupId: 'number',
            managedList: 'string',
            managedRulesets: { 'type': 'array', 'itemType': WafRuleConfigManagedRulesets_1.WafRuleConfigManagedRulesets },
            name: 'string',
            notes: 'string',
            rateLimit: WafRuleConfigRateLimit_1.WafRuleConfigRateLimit,
            securityLevel: WafRuleConfigSecurityLevel_1.WafRuleConfigSecurityLevel,
            sigchl: { 'type': 'array', 'itemType': 'string' },
            status: 'string',
            timer: WafTimer_1.WafTimer,
            type: 'string',
            value: 'string',
        };
    }
    validate() {
        if (this.actions && typeof this.actions.validate === 'function') {
            this.actions.validate();
        }
        if (this.appPackage && typeof this.appPackage.validate === 'function') {
            this.appPackage.validate();
        }
        if (this.appSdk && typeof this.appSdk.validate === 'function') {
            this.appSdk.validate();
        }
        if (Array.isArray(this.managedRulesets)) {
            $dara.Model.validateArray(this.managedRulesets);
        }
        if (this.rateLimit && typeof this.rateLimit.validate === 'function') {
            this.rateLimit.validate();
        }
        if (this.securityLevel && typeof this.securityLevel.validate === 'function') {
            this.securityLevel.validate();
        }
        if (Array.isArray(this.sigchl)) {
            $dara.Model.validateArray(this.sigchl);
        }
        if (this.timer && typeof this.timer.validate === 'function') {
            this.timer.validate();
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.WafRuleConfig = WafRuleConfig;
