// @alicloud/esa20240910/dist/models/WafSiteSettings.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WafSiteSettings = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const WafSiteSettingsAddBotProtectionHeaders_1 = require("./WafSiteSettingsAddBotProtectionHeaders");
const WafSiteSettingsAddSecurityHeaders_1 = require("./WafSiteSettingsAddSecurityHeaders");
const WafSiteSettingsBotManagement_1 = require("./WafSiteSettingsBotManagement");
const WafSiteSettingsClientIpIdentifier_1 = require("./WafSiteSettingsClientIpIdentifier");
const WafSiteSettingsSecurityLevel_1 = require("./WafSiteSettingsSecurityLevel");
class WafSiteSettings extends $dara.Model {
    static names() {
        return {
            addBotProtectionHeaders: 'AddBotProtectionHeaders',
            addSecurityHeaders: 'AddSecurityHeaders',
            botManagement: 'BotManagement',
            clientIpIdentifier: 'ClientIpIdentifier',
            securityLevel: 'SecurityLevel',
        };
    }
    static types() {
        return {
            addBotProtectionHeaders: WafSiteSettingsAddBotProtectionHeaders_1.WafSiteSettingsAddBotProtectionHeaders,
            addSecurityHeaders: WafSiteSettingsAddSecurityHeaders_1.WafSiteSettingsAddSecurityHeaders,
            botManagement: WafSiteSettingsBotManagement_1.WafSiteSettingsBotManagement,
            clientIpIdentifier: WafSiteSettingsClientIpIdentifier_1.WafSiteSettingsClientIpIdentifier,
            securityLevel: WafSiteSettingsSecurityLevel_1.WafSiteSettingsSecurityLevel,
        };
    }
    validate() {
        if (this.addBotProtectionHeaders && typeof this.addBotProtectionHeaders.validate === 'function') {
            this.addBotProtectionHeaders.validate();
        }
        if (this.addSecurityHeaders && typeof this.addSecurityHeaders.validate === 'function') {
            this.addSecurityHeaders.validate();
        }
        if (this.botManagement && typeof this.botManagement.validate === 'function') {
            this.botManagement.validate();
        }
        if (this.clientIpIdentifier && typeof this.clientIpIdentifier.validate === 'function') {
            this.clientIpIdentifier.validate();
        }
        if (this.securityLevel && typeof this.securityLevel.validate === 'function') {
            this.securityLevel.validate();
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.WafSiteSettings = WafSiteSettings;
