// @alicloud/esa20240910/dist/models/CreateSiteDeliveryTaskRequest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateSiteDeliveryTaskRequest = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const CreateSiteDeliveryTaskRequestHttpDelivery_1 = require("./CreateSiteDeliveryTaskRequestHttpDelivery");
const CreateSiteDeliveryTaskRequestKafkaDelivery_1 = require("./CreateSiteDeliveryTaskRequestKafkaDelivery");
const CreateSiteDeliveryTaskRequestOssDelivery_1 = require("./CreateSiteDeliveryTaskRequestOssDelivery");
const CreateSiteDeliveryTaskRequestS3delivery_1 = require("./CreateSiteDeliveryTaskRequestS3delivery");
const CreateSiteDeliveryTaskRequestSlsDelivery_1 = require("./CreateSiteDeliveryTaskRequestSlsDelivery");
class CreateSiteDeliveryTaskRequest extends $dara.Model {
    static names() {
        return {
            businessType: 'BusinessType',
            dataCenter: 'DataCenter',
            deliveryType: 'DeliveryType',
            discardRate: 'DiscardRate',
            fieldName: 'FieldName',
            httpDelivery: 'HttpDelivery',
            kafkaDelivery: 'KafkaDelivery',
            ossDelivery: 'OssDelivery',
            s3Delivery: 'S3Delivery',
            siteId: 'SiteId',
            slsDelivery: 'SlsDelivery',
            taskName: 'TaskName',
        };
    }
    static types() {
        return {
            businessType: 'string',
            dataCenter: 'string',
            deliveryType: 'string',
            discardRate: 'number',
            fieldName: 'string',
            httpDelivery: CreateSiteDeliveryTaskRequestHttpDelivery_1.CreateSiteDeliveryTaskRequestHttpDelivery,
            kafkaDelivery: CreateSiteDeliveryTaskRequestKafkaDelivery_1.CreateSiteDeliveryTaskRequestKafkaDelivery,
            ossDelivery: CreateSiteDeliveryTaskRequestOssDelivery_1.CreateSiteDeliveryTaskRequestOssDelivery,
            s3Delivery: CreateSiteDeliveryTaskRequestS3delivery_1.CreateSiteDeliveryTaskRequestS3Delivery,
            siteId: 'number',
            slsDelivery: CreateSiteDeliveryTaskRequestSlsDelivery_1.CreateSiteDeliveryTaskRequestSlsDelivery,
            taskName: 'string',
        };
    }
    validate() {
        if (this.httpDelivery && typeof this.httpDelivery.validate === 'function') {
            this.httpDelivery.validate();
        }
        if (this.kafkaDelivery && typeof this.kafkaDelivery.validate === 'function') {
            this.kafkaDelivery.validate();
        }
        if (this.ossDelivery && typeof this.ossDelivery.validate === 'function') {
            this.ossDelivery.validate();
        }
        if (this.s3Delivery && typeof this.s3Delivery.validate === 'function') {
            this.s3Delivery.validate();
        }
        if (this.slsDelivery && typeof this.slsDelivery.validate === 'function') {
            this.slsDelivery.validate();
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.CreateSiteDeliveryTaskRequest = CreateSiteDeliveryTaskRequest;
