// @alicloud/esa20240910/dist/models/CreateUserDeliveryTaskRequestHttpDelivery.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateUserDeliveryTaskRequestHttpDelivery = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const HttpDeliveryHeaderParamValue_1 = require("./HttpDeliveryHeaderParamValue");
const HttpDeliveryQueryParamValue_1 = require("./HttpDeliveryQueryParamValue");
const CreateUserDeliveryTaskRequestHttpDeliveryStandardAuthParam_1 = require("./CreateUserDeliveryTaskRequestHttpDeliveryStandardAuthParam");
class CreateUserDeliveryTaskRequestHttpDelivery extends $dara.Model {
    static names() {
        return {
            compress: 'Compress',
            destUrl: 'DestUrl',
            headerParam: 'HeaderParam',
            lastLogSplit: 'LastLogSplit',
            logBodyPrefix: 'LogBodyPrefix',
            logBodySuffix: 'LogBodySuffix',
            logSplit: 'LogSplit',
            logSplitWords: 'LogSplitWords',
            maxBackoffMS: 'MaxBackoffMS',
            maxBatchMB: 'MaxBatchMB',
            maxBatchSize: 'MaxBatchSize',
            maxRetry: 'MaxRetry',
            minBackoffMS: 'MinBackoffMS',
            queryParam: 'QueryParam',
            responseBodyKey: 'ResponseBodyKey',
            standardAuthOn: 'StandardAuthOn',
            standardAuthParam: 'StandardAuthParam',
            successCode: 'SuccessCode',
            transformTimeout: 'TransformTimeout',
        };
    }
    static types() {
        return {
            compress: 'string',
            destUrl: 'string',
            headerParam: { 'type': 'map', 'keyType': 'string', 'valueType': HttpDeliveryHeaderParamValue_1.HttpDeliveryHeaderParamValue },
            lastLogSplit: 'string',
            logBodyPrefix: 'string',
            logBodySuffix: 'string',
            logSplit: 'string',
            logSplitWords: 'string',
            maxBackoffMS: 'number',
            maxBatchMB: 'number',
            maxBatchSize: 'number',
            maxRetry: 'number',
            minBackoffMS: 'number',
            queryParam: { 'type': 'map', 'keyType': 'string', 'valueType': HttpDeliveryQueryParamValue_1.HttpDeliveryQueryParamValue },
            responseBodyKey: 'string',
            standardAuthOn: 'boolean',
            standardAuthParam: CreateUserDeliveryTaskRequestHttpDeliveryStandardAuthParam_1.CreateUserDeliveryTaskRequestHttpDeliveryStandardAuthParam,
            successCode: 'number',
            transformTimeout: 'number',
        };
    }
    validate() {
        if (this.headerParam) {
            $dara.Model.validateMap(this.headerParam);
        }
        if (this.queryParam) {
            $dara.Model.validateMap(this.queryParam);
        }
        if (this.standardAuthParam && typeof this.standardAuthParam.validate === 'function') {
            this.standardAuthParam.validate();
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.CreateUserDeliveryTaskRequestHttpDelivery = CreateUserDeliveryTaskRequestHttpDelivery;
