// @alicloud/esa20240910/dist/models/CreateLoadBalancerShrinkRequest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateLoadBalancerShrinkRequest = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
class CreateLoadBalancerShrinkRequest extends $dara.Model {
    static names() {
        return {
            adaptiveRoutingShrink: 'AdaptiveRouting',
            defaultPoolsShrink: 'DefaultPools',
            description: 'Description',
            enabled: 'Enabled',
            fallbackPool: 'FallbackPool',
            monitorShrink: 'Monitor',
            name: 'Name',
            randomSteeringShrink: 'RandomSteering',
            regionPools: 'RegionPools',
            rulesShrink: 'Rules',
            sessionAffinity: 'SessionAffinity',
            siteId: 'SiteId',
            steeringPolicy: 'SteeringPolicy',
            subRegionPools: 'SubRegionPools',
            ttl: 'Ttl',
        };
    }
    static types() {
        return {
            adaptiveRoutingShrink: 'string',
            defaultPoolsShrink: 'string',
            description: 'string',
            enabled: 'boolean',
            fallbackPool: 'number',
            monitorShrink: 'string',
            name: 'string',
            randomSteeringShrink: 'string',
            regionPools: 'any',
            rulesShrink: 'string',
            sessionAffinity: 'string',
            siteId: 'number',
            steeringPolicy: 'string',
            subRegionPools: 'any',
            ttl: 'number',
        };
    }
    validate() {
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.CreateLoadBalancerShrinkRequest = CreateLoadBalancerShrinkRequest;
