// @alicloud/esa20240910/dist/models/CreateEdgeContainerAppRequest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateEdgeContainerAppRequest = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
class CreateEdgeContainerAppRequest extends $dara.Model {
    static names() {
        return {
            healthCheckFailTimes: 'HealthCheckFailTimes',
            healthCheckHost: 'HealthCheckHost',
            healthCheckHttpCode: 'HealthCheckHttpCode',
            healthCheckInterval: 'HealthCheckInterval',
            healthCheckMethod: 'HealthCheckMethod',
            healthCheckPort: 'HealthCheckPort',
            healthCheckSuccTimes: 'HealthCheckSuccTimes',
            healthCheckTimeout: 'HealthCheckTimeout',
            healthCheckType: 'HealthCheckType',
            healthCheckURI: 'HealthCheckURI',
            name: 'Name',
            remarks: 'Remarks',
            servicePort: 'ServicePort',
            targetPort: 'TargetPort',
        };
    }
    static types() {
        return {
            healthCheckFailTimes: 'number',
            healthCheckHost: 'string',
            healthCheckHttpCode: 'string',
            healthCheckInterval: 'number',
            healthCheckMethod: 'string',
            healthCheckPort: 'number',
            healthCheckSuccTimes: 'number',
            healthCheckTimeout: 'number',
            healthCheckType: 'string',
            healthCheckURI: 'string',
            name: 'string',
            remarks: 'string',
            servicePort: 'number',
            targetPort: 'number',
        };
    }
    validate() {
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.CreateEdgeContainerAppRequest = CreateEdgeContainerAppRequest;
