// @alicloud/esa20240910/dist/models/CreateUserDeliveryTaskRequest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateUserDeliveryTaskRequest = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const CreateUserDeliveryTaskRequestHttpDelivery_1 = require("./CreateUserDeliveryTaskRequestHttpDelivery");
const CreateUserDeliveryTaskRequestKafkaDelivery_1 = require("./CreateUserDeliveryTaskRequestKafkaDelivery");
const CreateUserDeliveryTaskRequestOssDelivery_1 = require("./CreateUserDeliveryTaskRequestOssDelivery");
const CreateUserDeliveryTaskRequestS3delivery_1 = require("./CreateUserDeliveryTaskRequestS3delivery");
const CreateUserDeliveryTaskRequestSlsDelivery_1 = require("./CreateUserDeliveryTaskRequestSlsDelivery");
class CreateUserDeliveryTaskRequest extends $dara.Model {
    static names() {
        return {
            businessType: 'BusinessType',
            dataCenter: 'DataCenter',
            deliveryType: 'DeliveryType',
            details: 'Details',
            discardRate: 'DiscardRate',
            fieldName: 'FieldName',
            httpDelivery: 'HttpDelivery',
            kafkaDelivery: 'KafkaDelivery',
            ossDelivery: 'OssDelivery',
            s3Delivery: 'S3Delivery',
            slsDelivery: 'SlsDelivery',
            taskName: 'TaskName',
        };
    }
    static types() {
        return {
            businessType: 'string',
            dataCenter: 'string',
            deliveryType: 'string',
            details: 'string',
            discardRate: 'number',
            fieldName: 'string',
            httpDelivery: CreateUserDeliveryTaskRequestHttpDelivery_1.CreateUserDeliveryTaskRequestHttpDelivery,
            kafkaDelivery: CreateUserDeliveryTaskRequestKafkaDelivery_1.CreateUserDeliveryTaskRequestKafkaDelivery,
            ossDelivery: CreateUserDeliveryTaskRequestOssDelivery_1.CreateUserDeliveryTaskRequestOssDelivery,
            s3Delivery: CreateUserDeliveryTaskRequestS3delivery_1.CreateUserDeliveryTaskRequestS3Delivery,
            slsDelivery: CreateUserDeliveryTaskRequestSlsDelivery_1.CreateUserDeliveryTaskRequestSlsDelivery,
            taskName: 'string',
        };
    }
    validate() {
        if (this.httpDelivery && typeof this.httpDelivery.validate === 'function') {
            this.httpDelivery.validate();
        }
        if (this.kafkaDelivery && typeof this.kafkaDelivery.validate === 'function') {
            this.kafkaDelivery.validate();
        }
        if (this.ossDelivery && typeof this.ossDelivery.validate === 'function') {
            this.ossDelivery.validate();
        }
        if (this.s3Delivery && typeof this.s3Delivery.validate === 'function') {
            this.s3Delivery.validate();
        }
        if (this.slsDelivery && typeof this.slsDelivery.validate === 'function') {
            this.slsDelivery.validate();
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.CreateUserDeliveryTaskRequest = CreateUserDeliveryTaskRequest;
