// @alicloud/esa20240910/dist/models/GetEdgeContainerAppVersionResponseBodyVersionContainers.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetEdgeContainerAppVersionResponseBodyVersionContainers = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const GetEdgeContainerAppVersionResponseBodyVersionContainersAcrimageInfo_1 = require("./GetEdgeContainerAppVersionResponseBodyVersionContainersAcrimageInfo");
const GetEdgeContainerAppVersionResponseBodyVersionContainersProbeContent_1 = require("./GetEdgeContainerAppVersionResponseBodyVersionContainersProbeContent");
class GetEdgeContainerAppVersionResponseBodyVersionContainers extends $dara.Model {
    static names() {
        return {
            ACRImageInfo: 'ACRImageInfo',
            args: 'Args',
            command: 'Command',
            envVariables: 'EnvVariables',
            image: 'Image',
            isACRImage: 'IsACRImage',
            name: 'Name',
            postStart: 'PostStart',
            preStop: 'PreStop',
            probeContent: 'ProbeContent',
            probeType: 'ProbeType',
            spec: 'Spec',
            storage: 'Storage',
        };
    }
    static types() {
        return {
            ACRImageInfo: GetEdgeContainerAppVersionResponseBodyVersionContainersAcrimageInfo_1.GetEdgeContainerAppVersionResponseBodyVersionContainersACRImageInfo,
            args: 'string',
            command: 'string',
            envVariables: 'string',
            image: 'string',
            isACRImage: 'boolean',
            name: 'string',
            postStart: 'string',
            preStop: 'string',
            probeContent: GetEdgeContainerAppVersionResponseBodyVersionContainersProbeContent_1.GetEdgeContainerAppVersionResponseBodyVersionContainersProbeContent,
            probeType: 'string',
            spec: 'string',
            storage: 'string',
        };
    }
    validate() {
        if (this.ACRImageInfo && typeof this.ACRImageInfo.validate === 'function') {
            this.ACRImageInfo.validate();
        }
        if (this.probeContent && typeof this.probeContent.validate === 'function') {
            this.probeContent.validate();
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.GetEdgeContainerAppVersionResponseBodyVersionContainers = GetEdgeContainerAppVersionResponseBodyVersionContainers;
