// @alicloud/esa20240910/dist/models/GetOriginPoolResponseBodyReferences.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetOriginPoolResponseBodyReferences = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const GetOriginPoolResponseBodyReferencesDnsRecords_1 = require("./GetOriginPoolResponseBodyReferencesDnsRecords");
const GetOriginPoolResponseBodyReferencesIparecords_1 = require("./GetOriginPoolResponseBodyReferencesIparecords");
const GetOriginPoolResponseBodyReferencesLoadBalancers_1 = require("./GetOriginPoolResponseBodyReferencesLoadBalancers");
class GetOriginPoolResponseBodyReferences extends $dara.Model {
    static names() {
        return {
            dnsRecords: 'DnsRecords',
            IPARecords: 'IPARecords',
            loadBalancers: 'LoadBalancers',
        };
    }
    static types() {
        return {
            dnsRecords: { 'type': 'array', 'itemType': GetOriginPoolResponseBodyReferencesDnsRecords_1.GetOriginPoolResponseBodyReferencesDnsRecords },
            IPARecords: { 'type': 'array', 'itemType': GetOriginPoolResponseBodyReferencesIparecords_1.GetOriginPoolResponseBodyReferencesIPARecords },
            loadBalancers: { 'type': 'array', 'itemType': GetOriginPoolResponseBodyReferencesLoadBalancers_1.GetOriginPoolResponseBodyReferencesLoadBalancers },
        };
    }
    validate() {
        if (Array.isArray(this.dnsRecords)) {
            $dara.Model.validateArray(this.dnsRecords);
        }
        if (Array.isArray(this.IPARecords)) {
            $dara.Model.validateArray(this.IPARecords);
        }
        if (Array.isArray(this.loadBalancers)) {
            $dara.Model.validateArray(this.loadBalancers);
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.GetOriginPoolResponseBodyReferences = GetOriginPoolResponseBodyReferences;
