// @alicloud/esa20240910/dist/models/GetWafQuotaResponseBodyQuota.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetWafQuotaResponseBodyQuota = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const GetWafQuotaResponseBodyQuotaList_1 = require("./GetWafQuotaResponseBodyQuotaList");
const GetWafQuotaResponseBodyQuotaManagedRulesGroup_1 = require("./GetWafQuotaResponseBodyQuotaManagedRulesGroup");
const GetWafQuotaResponseBodyQuotaPage_1 = require("./GetWafQuotaResponseBodyQuotaPage");
const GetWafQuotaResponseBodyQuotaScenePolicy_1 = require("./GetWafQuotaResponseBodyQuotaScenePolicy");
class GetWafQuotaResponseBodyQuota extends $dara.Model {
    static names() {
        return {
            list: 'List',
            managedRulesGroup: 'ManagedRulesGroup',
            page: 'Page',
            scenePolicy: 'ScenePolicy',
        };
    }
    static types() {
        return {
            list: GetWafQuotaResponseBodyQuotaList_1.GetWafQuotaResponseBodyQuotaList,
            managedRulesGroup: GetWafQuotaResponseBodyQuotaManagedRulesGroup_1.GetWafQuotaResponseBodyQuotaManagedRulesGroup,
            page: GetWafQuotaResponseBodyQuotaPage_1.GetWafQuotaResponseBodyQuotaPage,
            scenePolicy: GetWafQuotaResponseBodyQuotaScenePolicy_1.GetWafQuotaResponseBodyQuotaScenePolicy,
        };
    }
    validate() {
        if (this.list && typeof this.list.validate === 'function') {
            this.list.validate();
        }
        if (this.managedRulesGroup && typeof this.managedRulesGroup.validate === 'function') {
            this.managedRulesGroup.validate();
        }
        if (this.page && typeof this.page.validate === 'function') {
            this.page.validate();
        }
        if (this.scenePolicy && typeof this.scenePolicy.validate === 'function') {
            this.scenePolicy.validate();
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.GetWafQuotaResponseBodyQuota = GetWafQuotaResponseBodyQuota;
