// @alicloud/esa20240910/dist/models/UpdateCacheRuleRequest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateCacheRuleRequest = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
class UpdateCacheRuleRequest extends $dara.Model {
    static names() {
        return {
            additionalCacheablePorts: 'AdditionalCacheablePorts',
            browserCacheMode: 'BrowserCacheMode',
            browserCacheTtl: 'BrowserCacheTtl',
            bypassCache: 'BypassCache',
            cacheDeceptionArmor: 'CacheDeceptionArmor',
            cacheReserveEligibility: 'CacheReserveEligibility',
            checkPresenceCookie: 'CheckPresenceCookie',
            checkPresenceHeader: 'CheckPresenceHeader',
            configId: 'ConfigId',
            edgeCacheMode: 'EdgeCacheMode',
            edgeCacheTtl: 'EdgeCacheTtl',
            edgeStatusCodeCacheTtl: 'EdgeStatusCodeCacheTtl',
            includeCookie: 'IncludeCookie',
            includeHeader: 'IncludeHeader',
            queryString: 'QueryString',
            queryStringMode: 'QueryStringMode',
            rule: 'Rule',
            ruleEnable: 'RuleEnable',
            ruleName: 'RuleName',
            sequence: 'Sequence',
            serveStale: 'ServeStale',
            siteId: 'SiteId',
            sortQueryStringForCache: 'SortQueryStringForCache',
            userDeviceType: 'UserDeviceType',
            userGeo: 'UserGeo',
            userLanguage: 'UserLanguage',
        };
    }
    static types() {
        return {
            additionalCacheablePorts: 'string',
            browserCacheMode: 'string',
            browserCacheTtl: 'string',
            bypassCache: 'string',
            cacheDeceptionArmor: 'string',
            cacheReserveEligibility: 'string',
            checkPresenceCookie: 'string',
            checkPresenceHeader: 'string',
            configId: 'number',
            edgeCacheMode: 'string',
            edgeCacheTtl: 'string',
            edgeStatusCodeCacheTtl: 'string',
            includeCookie: 'string',
            includeHeader: 'string',
            queryString: 'string',
            queryStringMode: 'string',
            rule: 'string',
            ruleEnable: 'string',
            ruleName: 'string',
            sequence: 'number',
            serveStale: 'string',
            siteId: 'number',
            sortQueryStringForCache: 'string',
            userDeviceType: 'string',
            userGeo: 'string',
            userLanguage: 'string',
        };
    }
    validate() {
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.UpdateCacheRuleRequest = UpdateCacheRuleRequest;
