// @alicloud/esa20240910/dist/models/WafSiteSettingsBotManagement.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WafSiteSettingsBotManagement = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const WafSiteSettingsBotManagementDefiniteBots_1 = require("./WafSiteSettingsBotManagementDefiniteBots");
const WafSiteSettingsBotManagementEffectOnStatic_1 = require("./WafSiteSettingsBotManagementEffectOnStatic");
const WafSiteSettingsBotManagementJsdetection_1 = require("./WafSiteSettingsBotManagementJsdetection");
const WafSiteSettingsBotManagementLikelyBots_1 = require("./WafSiteSettingsBotManagementLikelyBots");
const WafSiteSettingsBotManagementVerifiedBots_1 = require("./WafSiteSettingsBotManagementVerifiedBots");
class WafSiteSettingsBotManagement extends $dara.Model {
    static names() {
        return {
            definiteBots: 'DefiniteBots',
            effectOnStatic: 'EffectOnStatic',
            JSDetection: 'JSDetection',
            likelyBots: 'LikelyBots',
            verifiedBots: 'VerifiedBots',
        };
    }
    static types() {
        return {
            definiteBots: WafSiteSettingsBotManagementDefiniteBots_1.WafSiteSettingsBotManagementDefiniteBots,
            effectOnStatic: WafSiteSettingsBotManagementEffectOnStatic_1.WafSiteSettingsBotManagementEffectOnStatic,
            JSDetection: WafSiteSettingsBotManagementJsdetection_1.WafSiteSettingsBotManagementJSDetection,
            likelyBots: WafSiteSettingsBotManagementLikelyBots_1.WafSiteSettingsBotManagementLikelyBots,
            verifiedBots: WafSiteSettingsBotManagementVerifiedBots_1.WafSiteSettingsBotManagementVerifiedBots,
        };
    }
    validate() {
        if (this.definiteBots && typeof this.definiteBots.validate === 'function') {
            this.definiteBots.validate();
        }
        if (this.effectOnStatic && typeof this.effectOnStatic.validate === 'function') {
            this.effectOnStatic.validate();
        }
        if (this.JSDetection && typeof this.JSDetection.validate === 'function') {
            this.JSDetection.validate();
        }
        if (this.likelyBots && typeof this.likelyBots.validate === 'function') {
            this.likelyBots.validate();
        }
        if (this.verifiedBots && typeof this.verifiedBots.validate === 'function') {
            this.verifiedBots.validate();
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.WafSiteSettingsBotManagement = WafSiteSettingsBotManagement;
